### PR TITLE
Modified rerouting logic so we also apply a similar slower route for when a trafficjam occurs

### PR DESCRIPTION
**Pull request by KasparElmans**
# Description
Currently we only apply the route when the route is _faster_ (that is what the logic is used for), so the same route with a slower ETA is not applied. 

In this PR I've modified the rerouting logic, so we can display more accurate ETA's within the client app. This is especially important when the user drives towards/within a traffic jam. 

## Rerouting logic:
Every 2 minutes we ask for a new route.

## When do we apply a new route?
_Before:_
- Check if new route is 10% or more faster than the current route.
    - If so -> Apply the new route

_After:_
- Check if the new route is 10% or more faster than the current route. 
    - If so -> Apply the new route
- If slower -> Check if the route is 90% or more similar to current route (by geometry)
    - If so -> Check if new ETA is more than 30s different than current route
        - If so -> Apply the new route
        
**Regarding geometry check:**
We assume the geometry from the routing endpoint is _stable_. It uses simple coordinate comparison checks (optimising for performance). 

### Summary of changes
- Added method to compare route geometries that gives a percentage of the match
- Added a 'best match' method for rerouting that finds the most similar route that is >90% similar 
- Made test targets run again after FM changes & migrating to Maplibre by updating dependencies (Quick & Nimble) and fixing compiler errors in those targets
- Significantly refactored `RouteController`'s rerouting logic, now returning a 'reroute reason' when `didRerouteAlong` delegate gets called, so that API is changed in signature
- Moved & renamed `RouteControllerProactiveReroutingInterval` to be an instance property of `RouteController` so the user can easily set this per route controller 
- Added unit tests for new matching & rerouting methods
- Added `shouldCheckForRerouteInLastMinutes` property to `RouteController` to skip check if we have 10+ mins left in the route, to also fetch routes for an ETA update in the last 10 mins of the route

### How to test
Run the new tests for yourself and check if they are valid

### CODE REVIEW

File: Cartfile [hunk 1]
```diff
@@ -1,6 +1,6 @@
 github "mapbox/MapboxGeocoder.swift" ~> 0.10
 github "mapbox/mapbox-events-ios" "v0.10.2"
-github "mapbox/MapboxDirections.swift" ~> 0.23.0
+github "flitsmeister/mapbox-directions-swift" "feature/test"
```
Comment (KasparElmans) [inline Cartfile]: Dit moeten we nog resetten.

File: MapboxCoreNavigation/RouteController.swift [hunk 2]
```diff
@@ -426,7 +423,9 @@ extension RouteController: CLLocationManagerDelegate {
         return false
     }
 
-    func checkForFasterRoute(from location: CLLocation) {
+    public static var isMegaFileActive = false
```
Comment (KasparElmans) [inline MapboxCoreNavigation/RouteController.swift]: Delete?

File: MapboxCoreNavigation/RouteController.swift [hunk 3]
```diff
@@ -441,43 +440,86 @@ extension RouteController: CLLocationManagerDelegate {
         }
 
         // Only check every so often for a faster route.
-        guard location.timestamp.timeIntervalSince(lastLocationDate) >= RouteControllerProactiveReroutingInterval else {
-            return
-        }
+//        guard location.timestamp.timeIntervalSince(lastLocationDate) >= RouteControllerProactiveReroutingInterval else {
+//            return
+//        }
+
```
Comment (KasparElmans) [inline MapboxCoreNavigation/RouteController.swift]: Reset this

File: MapboxCoreNavigation/Constants.swift [hunk 4]
```diff
@@ -76,7 +76,7 @@ public var RouteControllerNumberOfSecondsForRerouteFeedback: TimeInterval = 10
 /**
  The number of seconds between attempts to automatically calculate a more optimal route while traveling.
  */
-public var RouteControllerProactiveReroutingInterval: TimeInterval = 120
+public var RouteControllerProactiveReroutingInterval: TimeInterval = 10
```
Comment (KasparElmans) [inline MapboxCoreNavigation/Constants.swift]: Dit moeten we terugzetten.

File: MapboxCoreNavigation/RouteController.swift [hunk 5]
```diff
@@ -13,6 +13,11 @@ import UIKit
  */
 @objc(MBRouteController)
 open class RouteController: NSObject, Router {
+    
+    /**
+     The number of seconds between attempts to automatically calculate a more optimal route while traveling.
+     */
+    public static var RouteControllerProactiveReroutingInterval: TimeInterval = 120
```
Comment (Svantulden) [inline MapboxCoreNavigation/RouteController.swift]: Moved so the client can override this

File: MapboxCoreNavigation/RouteController.swift [hunk 6]
```diff
@@ -441,48 +455,76 @@ extension RouteController: CLLocationManagerDelegate {
...
-        guard location.timestamp.timeIntervalSince(lastLocationDate) >= RouteControllerProactiveReroutingInterval else {
+        guard location.timestamp.timeIntervalSince(lastLocationDate) >= Self.RouteControllerProactiveReroutingInterval else {
             return
         }
+        
         let durationRemaining = routeProgress.durationRemaining
         
         isFindingFasterRoute = true
+        
+        // TODO: Remove debug statements
```
Comment (Svantulden) [inline MapboxCoreNavigation/RouteController.swift]: Remove TODO's before merging

File: MapboxCoreNavigationTests/RouteControllerTests.swift [hunk 7]
```diff
@@ -360,12 +317,159 @@ class RouteControllerTests: XCTestCase {
...
+    lazy var a12ToVeenendaalBiggerDetour = {
+        let waypoint1 = Waypoint(coordinate: CLLocationCoordinate2D(latitude: 52.02224357, longitude: 5.78149084))
+        let waypoint2 = Waypoint(coordinate: CLLocationCoordinate2D(latitude: 52.04520875, longitude: 5.5748937))
+        let waypoint3 = Waypoint(coordinate: CLLocationCoordinate2D(latitude: 52.03924958, longitude: 5.55054131))
+        let route = Route(json: Fixture.JSONFromFileNamed(name: "A12-To-Veenendaal-Bigger-Detour"), waypoints: [waypoint1, waypoint2, waypoint3], options: NavigationRouteOptions(waypoints: [waypoint1, waypoint2, waypoint3]))
+        route.accessToken = Constants.accessToken
+        return route
+    }()
+
+    func testRouteControllerMatchPercentage() {
```
Comment (Svantulden) [inline MapboxCoreNavigationTests/RouteControllerTests.swift:433]: From this line on we have new tests that need review

File: MapboxCoreNavigation/RouteController.swift [hunk 8]
```diff
@@ -475,50 +475,78 @@ extension RouteController: CLLocationManagerDelegate {
...
             }
             
-            // Check if the new route is faster by comparing the ETA to the current ETA. Should be 10% faster or more
-            let routeIsFaster = firstStep.expectedTravelTime >= RouteControllerMediumAlertInterval && currentUpcomingManeuver == firstLeg.steps[1] && route.expectedTravelTime <= 0.9 * durationRemaining
-            
+            self.applyNewRerouteIfNeeded(mostSimilarRoute: route, allRoutes: routes, currentUpcomingManeuver: currentUpcomingManeuver, durationRemaining: durationRemaining)
+        }
+    }
+    
+    func applyNewRerouteIfNeeded(mostSimilarRoute: Route, allRoutes: [Route], currentUpcomingManeuver: RouteStep, durationRemaining: TimeInterval) {
```
Comment (Svantulden) [inline MapboxCoreNavigation/RouteController.swift:540]: New abstracted function that should be reviewed critically

File: MapboxCoreNavigation/RouteController.swift [hunk 9]
```diff
@@ -334,19 +389,16 @@ extension RouteController: CLLocationManagerDelegate {
...
 
         // Check for faster route given users current location
         guard reroutesProactively else { return }
-        // Only check for faster alternatives if the user has plenty of time left on the route.
-        guard routeProgress.durationRemaining > 600 else { return }
-        // If the user is approaching a maneuver, don't check for a faster alternatives
-        guard routeProgress.currentLegProgress.currentStepProgress.durationRemaining > RouteControllerMediumAlertInterval else { return }
-        checkForFasterRoute(from: location)
+        
+        checkForNewRoute(from: location)
```
Comment (Svantulden) [inline MapboxCoreNavigation/RouteController.swift]: Does removing these checks cause more backend calls and is this justified?

File: MapboxCoreNavigation/RouteProgress.swift [hunk 10]
```diff
@@ -512,7 +512,10 @@ open class RouteStepProgress: NSObject {
      An `Array` of remaining `SpokenInstruction` for a step.
      */
     @objc public var remainingSpokenInstructions: [SpokenInstruction]? {
-        guard let instructions = step.instructionsSpokenAlongStep else { return nil }
+        guard
+            let instructions = step.instructionsSpokenAlongStep,
+            spokenInstructionIndex <= instructions.endIndex
```
Comment (Svantulden) [inline MapboxCoreNavigation/RouteProgress.swift:517]: Why is this added? @KasparElmans

File: MapboxCoreNavigation/RouteController.swift [hunk 11]
```diff
@@ -67,6 +76,28 @@ open class RouteController: NSObject, Router {
             movementsAwayFromRoute = 0
         }
     }
+    
+    // TODO: Remove testing Route for debug
+    private lazy var a12ToVeenendaalNormalWithTraffic = {
```
Comment (Svantulden) [inline MapboxCoreNavigation/RouteController.swift]: Make helper method for this initialisation

File: MapboxCoreNavigation/RouteProgress.swift (hunk 10 above)
Comment (Svantulden) [reply MapboxCoreNavigation/RouteProgress.swift:517]: Can crash maybe

File: MapboxCoreNavigation/Route.swift [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+
+extension Route {
+    convenience init(jsonFileName: String, waypoints: [CLLocationCoordinate2D], polylineShapeFormat: RouteShapeFormat = .polyline6, bundle: Bundle = .main) {
+        let convertedWaypoints = waypoints.compactMap { waypoint in
+            Waypoint(coordinate: waypoint)
+        }
+        let routeOptions = NavigationRouteOptions(waypoints: convertedWaypoints)
+        routeOptions.shapeFormat = polylineShapeFormat
+        self.init(json: Fixture.JSONFromFileNamed(name: jsonFileName, bundle: bundle), waypoints: convertedWaypoints, options: routeOptions)
+        accessToken = "nonsense"
```
Comment (milotolboom) [inline MapboxCoreNavigation/Route.swift]: Setting the accessToken to `"nonsense"` seems a bit nonsensical 😄

File: MapboxCoreNavigation/RouteController.swift [hunk 13]
```diff
@@ -87,6 +97,30 @@ open class RouteController: NSObject, Router {
 
     var userSnapToStepDistanceFromManeuver: CLLocationDistance?
     
+    /// Describes a reason for rerouting and applying a new route
+    @objc public enum RerouteReason: Int, CustomStringConvertible {
+        /// When we check for a faster route we can also reroute the user when we just want to update the ETA. For example when the user is driving on a route where a Trafficjam appears, it should update the ETA
+        case ETAUpdate
```
Comment (milotolboom) [inline MapboxCoreNavigation/RouteController.swift:106]: Correct me if I'm wrong but this isn't strictly a reroute reason? This is a reason why the route gets updated but not re-routed

File: MapboxCoreNavigation/RouteController.swift [hunk 14]
```diff
@@ -441,48 +502,109 @@ extension RouteController: CLLocationManagerDelegate {
...
-        guard location.timestamp.timeIntervalSince(lastLocationDate) >= RouteControllerProactiveReroutingInterval else {
+        guard location.timestamp.timeIntervalSince(lastLocationDate) >= routeControllerProactiveReroutingInterval else {
             return
         }
+        
         let durationRemaining = routeProgress.durationRemaining
         
         isFindingFasterRoute = true
+        
+        print("[RouteController] Checking for faster/updated route...")
```
Comment (milotolboom) [inline MapboxCoreNavigation/RouteController.swift:519]: Do we use prints in this project?

File: MapboxNavigationTests/CarPlayManagerTests.swift [hunk 15]
```diff
@@ -1,320 +0,0 @@
-import XCTest
-import MapboxNavigation
-import MapboxCoreNavigation
-import MapboxDirections
```
Comment (milotolboom) [inline MapboxNavigationTests/CarPlayManagerTests.swift:4]: Why delete these tests?

File: MapboxCoreNavigation/RouteController.swift [hunk 16]
```diff
@@ -441,48 +502,109 @@ extension RouteController: CLLocationManagerDelegate {
...
+                print("[RouteController] Found matching route \(matchingRoute.matchPercentage)%, updating ETA...")
+                print("[RouteController] Duration remaining CURRENT: \(routeProgress.durationRemaining)")
+                print("[RouteController] Expected travel time: \(matchingRoute.route.expectedTravelTime)")
+                print("[RouteController] Set the new route")
+                
+                // Don't announce new route
+                routeProgress = RouteProgress(route: routeToApply, legIndex: 0, spokenInstructionIndex: routeProgress.currentLegProgress.currentStepProgress.spokenInstructionIndex)
+                
+                // Inform delegate
+                delegate?.routeController?(self, didRerouteAlong: routeToApply, reason: .ETAUpdate)
```
Comment (milotolboom) [inline MapboxCoreNavigation/RouteController.swift:607]: A slower route can still be a reroute right? Why is it marked as ETA update?

File: MapboxCoreNavigation/RouteController.swift [hunk 17]
```diff
@@ -694,8 +815,49 @@ extension RouteController: CLLocationManagerDelegate {
...
+    /// - Parameters:
+    ///   - route: First route to compare
+    ///   - route2: Second route to compare
+    /// - Returns: A percentage of the match. Can return `nil` if one of the route's geometry cannot be found.
+    static func matchPercentage(between route: Route, and route2: Route) -> Double? {
+        guard let currentRouteCoordinates = route.coordinates else { return nil }
+        guard let otherRouteCoordinates = route2.coordinates else { return nil }
+        
+        // Convert to strings, only taking 4 decimals
+        let currentRouteCoordinatesStrings = currentRouteCoordinates.map { String(format: "%.4f,%.4f", $0.latitude, $0.longitude) }
```
Comment (milotolboom) [inline MapboxCoreNavigation/RouteController.swift:835]: How was the 4 floating decimals chosen? Is this the desired accuracy?

**Review by milotolboom — APPROVED** [review]
Look good, just a few questions!

File: MapboxCoreNavigation/Route.swift (hunk 12 above)
Comment (Svantulden) [reply MapboxCoreNavigation/Route.swift]: TODO: Param

File: MapboxCoreNavigation/RouteController.swift [hunk 18]
```diff
@@ -13,6 +13,16 @@ import UIKit
...
+    
+    /**
+     The number of seconds between attempts to automatically calculate a more optimal route while traveling.
+     */
+    public var routeControllerProactiveReroutingInterval: TimeInterval = 120
+    
+    /// With this property you can enable test-routes to be returned when rerouting for an ETA update. It will randomly choose a route between two that differ a lot by ETA when rerouting
+    /// These will only get returned when that route is _slower_ than the current route, as that forces an ETA update reroute.
+    /// To pass the rerouting checks, simulate or drive a route from coordinate `52.02224357,5.78149084` to `52.03924958,5.55054131`
+    public var shouldReturnTestingETAUpdateReroutes: Bool = false
```
Comment (MaartenZonneveld) [inline MapboxCoreNavigation/RouteController.swift]:
```suggestion
    public var shouldReturnTestingETAUpdateReroutes = false
```

File: MapboxCoreNavigation/RouteController.swift [hunk 19]
```diff
@@ -334,19 +392,22 @@ extension RouteController: CLLocationManagerDelegate {
...
         }
 
         updateSpokenInstructionProgress()
 
         // Check for faster route given users current location
         guard reroutesProactively else { return }
-        // Only check for faster alternatives if the user has plenty of time left on the route.
+        
+        // Only check for faster routes or ETA updates if the user has plenty of time left on the route (10+min)
         guard routeProgress.durationRemaining > 600 else { return }
```
Comment (MaartenZonneveld) [inline MapboxCoreNavigation/RouteController.swift]: Why wouldn't we check for ETA updates in the last 10 minutes? I get you don't want to reroute anymore, but updated ETA is still relevant, as the ETA could theoretically increase significantly in those last 10 minutes.

File: MapboxCoreNavigation/RouteController.swift [hunk 20]
```diff
@@ -334,19 +392,22 @@ extension RouteController: CLLocationManagerDelegate {
...
-        // If the user is approaching a maneuver, don't check for a faster alternatives
+        
+        // If the user is approaching a maneuver (within 70secs of the maneuver), don't check for a faster routes or ETA updates
         guard routeProgress.currentLegProgress.currentStepProgress.durationRemaining > RouteControllerMediumAlertInterval else { return }
```
Comment (MaartenZonneveld) [inline MapboxCoreNavigation/RouteController.swift:414]: Also, why no ETA updates in this case? Again, for reroutes it makes sense, but I don't get why you wouldn't want to update ETA.

File: MapboxCoreNavigation/RouteController.swift (hunk 8 above)
Comment (MaartenZonneveld) [reply MapboxCoreNavigation/RouteController.swift:540]: Can this function have private access?

File: MapboxCoreNavigation/RouteController.swift [hunk 21]
```diff
@@ -441,48 +502,109 @@ extension RouteController: CLLocationManagerDelegate {
...
+    func applyNewRerouteIfNeeded(mostSimilarRoute: Route, allRoutes: [Route], currentUpcomingManeuver: RouteStep, durationRemaining: TimeInterval) {
+        guard let firstLeg = mostSimilarRoute.legs.first, let firstStep = firstLeg.steps.first else {
+            return
+        }
+        
+        // Current and First step of old and new route should be significant enough of a maneuver before applying a faster route, so we don't apply the route just before a maneuver will occur
+        let significantFirstStep = firstStep.expectedTravelTime >= RouteControllerMediumAlertInterval && routeProgress.currentLegProgress.currentStepProgress.durationRemaining > RouteControllerMediumAlertInterval
+        
+        // Current maneuver should correspond to the next maneuver in the new route
+        let sameUpcomingManeuver = currentUpcomingManeuver == firstLeg.steps[1]
```
Comment (MaartenZonneveld) [inline MapboxCoreNavigation/RouteController.swift]: Are you sure there are always at least 2 steps in a leg? Otherwise this array will go out of bounds and crash.

Comment (MaartenZonneveld) [reply MapboxCoreNavigation/RouteController.swift]: If you are sure, please comment :)

File: MapboxCoreNavigation/RouteController.swift [hunk 22]
```diff
@@ -441,48 +502,109 @@ extension RouteController: CLLocationManagerDelegate {
...
+        }
+    }
+    
+    func applyNewRerouteIfNeeded(mostSimilarRoute: Route, allRoutes: [Route], currentUpcomingManeuver: RouteStep, durationRemaining: TimeInterval) {
+        guard let firstLeg = mostSimilarRoute.legs.first, let firstStep = firstLeg.steps.first else {
+            return
+        }
+        
+        // Current and First step of old and new route should be significant enough of a maneuver before applying a faster route, so we don't apply the route just before a maneuver will occur
+        let significantFirstStep = firstStep.expectedTravelTime >= RouteControllerMediumAlertInterval && routeProgress.currentLegProgress.currentStepProgress.durationRemaining > RouteControllerMediumAlertInterval
```
Comment (MaartenZonneveld) [inline MapboxCoreNavigation/RouteController.swift]: At first I thought this let contained some Step object, but it is a boolean. Please consider the naming conventions for booleans:
```suggestion
        let isFirstStepSignificant = firstStep.expectedTravelTime >= RouteControllerMediumAlertInterval && routeProgress.currentLegProgress.currentStepProgress.durationRemaining > RouteControllerMediumAlertInterval
```

Comment (MaartenZonneveld) [reply MapboxCoreNavigation/RouteController.swift]: This applies to the naming of other booleans as well.

File: MapboxCoreNavigation/RouteController.swift [hunk 23]
```diff
@@ -441,48 +502,109 @@ extension RouteController: CLLocationManagerDelegate {
...
+        
+        // Current maneuver should correspond to the next maneuver in the new route
+        let sameUpcomingManeuver = currentUpcomingManeuver == firstLeg.steps[1]
+        
+        // Check if the new route is faster by comparing the ETA to the current ETA. Should be 10% faster or more
+        let routeIsFaster = mostSimilarRoute.expectedTravelTime <= 0.9 * durationRemaining
+        
+        // Only check for alternatives if the user has plenty of time left on the route (10min+)
+        let userHasEnoughTimeOnRoute = self.routeProgress.durationRemaining > 600
```
Comment (MaartenZonneveld) [inline MapboxCoreNavigation/RouteController.swift:555]:
```suggestion
        let userHasEnoughTimeOnRoute = routeProgress.durationRemaining > 600
```

Comment (MaartenZonneveld) [reply MapboxCoreNavigation/RouteController.swift:555]: This was already checked before, right?

File: MapboxCoreNavigation/RouteController.swift (hunk 13 above)
Comment (Svantulden) [reply MapboxCoreNavigation/RouteController.swift:106]: We see a reroute as the route being updated to a new model. When rerouting for an ETA update, the route can be (max) 10% different in geometry and have a different ETA

File: MapboxCoreNavigation/RouteController.swift (hunk 14 above)
Comment (Svantulden) [reply MapboxCoreNavigation/RouteController.swift:519]: For now we want to add clarity to the rerouting logic with these prints. We want to add robust logging functionality (sending logs with levels to the client) in the future

File: MapboxNavigationTests/CarPlayManagerTests.swift (hunk 15 above)
Comment (Svantulden) [reply MapboxNavigationTests/CarPlayManagerTests.swift:4]: Because Telemetry is cut out of the project, those tests don't work anymore and aren't relevant. The tests weren't updated yet. 

Also, in the newest Swift version we cannot subclass `CPInterfaceController`, so mocking doesn't make sense here anymore

File: MapboxCoreNavigation/RouteController.swift (hunk 16 above)
Comment (Svantulden) [reply MapboxCoreNavigation/RouteController.swift:607]: No, it has passed a check that the geometry is 90% or more the same so it 'just' updates the route with a new ETA

File: MapboxCoreNavigation/RouteController.swift (hunk 17 above)
Comment (Svantulden) [reply MapboxCoreNavigation/RouteController.swift:835]: @MaartenZonneveld chose this, I think to account to the small variability in routing responses where coordinates match but differ in the last few decimals.

File: MapboxCoreNavigation/RouteController.swift (hunk 19 above)
Comment (Svantulden) [reply MapboxCoreNavigation/RouteController.swift]: Backend was strict about not supporting more calls than we currently do client-side. We could configure this in the SDK so other clients can skip this check? What do you think?

File: MapboxCoreNavigation/RouteController.swift (hunk 20 above)
Comment (Svantulden) [reply MapboxCoreNavigation/RouteController.swift:414]: The route differs a maximum of 10%. That could differ theoretically in an instruction in the beginning of the new route so the user has to suddenly change direction. I've kept these checks to keep the change to a minimum before further examining which parts we can optimise to deliver a better experience with ETA updates. 

Do you think this is excessive?

File: MapboxCoreNavigation/RouteController.swift (hunk 8 above)
Comment (Svantulden) [reply MapboxCoreNavigation/RouteController.swift:540]: No, these are used in unit tests. Do you have a suggestion how to only make access public for testing?

File: MapboxCoreNavigation/RouteController.swift (hunk 21 above)
Comment (Svantulden) [reply MapboxCoreNavigation/RouteController.swift]: This behaviour isn't changed from the existing rerouting logic and I think routes always have a 'start' maneuver at the beginning. Not sure though, I can make this more safe

File: MapboxCoreNavigation/RouteController.swift (hunk 22 above)
Comment (Svantulden) [reply MapboxCoreNavigation/RouteController.swift]: Sure

File: MapboxCoreNavigation/RouteController.swift (hunk 23 above)
Comment (Svantulden) [reply MapboxCoreNavigation/RouteController.swift:555]: Theoretically this could change when a reroute call times out, but I'm fine with changing it 👍

File: MapboxCoreNavigation/RouteController.swift (hunk 19 above)
Comment (MaartenZonneveld) [reply MapboxCoreNavigation/RouteController.swift]: If it is a client requirement, we should make it configurable by the client

File: MapboxCoreNavigation/RouteController.swift (hunk 20 above)
Comment (MaartenZonneveld) [reply MapboxCoreNavigation/RouteController.swift:414]: I think that with your explanation it makes more sense. Let's see how this works IRL. We may want to tweak this later

**Review by MaartenZonneveld — APPROVED** [review]
